### DISCUSSION TITLE
3.7.0 Goals/Contents

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#962] [General]
What do we want to see in 3.7.0 ?

**Comment by bevans2000** [comment]
We have done a lot of internal stuff in 3.6.0 so we need new features in 3.7.0

**Comment by mokun** [comment]
@bevans2000 @bdowrick @Urwumpe @scud1 

Yes we could wrap up 3.6.0 this Sunday 9 Jul 2023. We've done at least two feature overhaul already and many minor improvement with 441 commits.

Feel free to chime in what you would like to work on or see in 3.7.0.

**Reply by Urwumpe** [reply]
I think about adding some graphics to it, like SpaceX themed landers. Maybe also a few more meals and recipes.

But my long-term goal would be allowing to customize rovers, maybe we could allow a special project in 3.7.0 already to replace  components, analog to the construction mission. But I have no idea how to realize it.

A maybe less far-sketched idea: Allow the settlers to have personal belongings, to make them more human-like.

**Reply by bevans2000** [reply]
I can't promise to finish my 2 remaining tasks by Sunday but I'll give it a go.

**Reply by mokun** [reply]
> customize rovers, maybe we could allow a special project in 3.7.0 already to replace components

May be letting players customize their own rovers ? 

That should include choosing propulsion system (or fuel cell stack, fuel tank size), equipment package, # of seats, # of motors, # of battery modules and capacity, etc.

So are you putting together a RPS rover just for fun ?

By the way, is it the sentiment of most Germans these days that nuclear reactors are no good and thus no possibly of developing a nuclear-powered rover then :D ?

**Reply by bevans2000** [reply]
This is very possible. The approach would be to let the user create their own ```VehicleSpec``` instance which should contain anything that can be customized. Logic already supports multiple VehicleSpec.
Also I would drop the different Rover types as well; it's not really used for anything. Let the parameters of the VehicleSpec drive vehicle selection.

**Comment by Urwumpe** [comment]
> So are you putting together a RPS rover just for fun ?

I thought more about lowering it a bit to one inch of ground clearance, install sewer-sized end pipes, paint stripes on the side and hang a fox tail on the antenna.

> 
> By the way, is it the sentiment of most Germans these days that nuclear reactors are no good and thus no possibly of developing a nuclear-powered rover then :D ?

That derails it a bit. The sentiment is a flag in the wind, for most of the time, its over 80% against and last winter it suddenly went to 69% pro, because some media and nazi party told lies about a looming blackout. Which, of course, never came. Today, those still point at the 3/5th of a nuclear power plants output, that we import from France (0.5-0.7 TWh/month) and claim, we depend on it (In reality, it just saves us some Euros, since we can shutdown more lignite power plants, the fossile fuel consumption in the electricity sector dropped by nearly 40%, if you compare this years June to the past years). Yeah, sorry, I'm an economic critic of nuclear power. Instead of maturing and getting cheaper and more reliable, the technology got more costly, has larger cost overruns additionally to it and its operators still manage to prepare the next large accident by accumulating technical debts. And we all knew about this trend this since 1969, and instead of getting this trend arrested, we decided to lie about it and report things to be better than they actually are.

So, sorry for the derailing accident. 

The nuclear reactor rover is simply a matter of size, weight and volume. We can't cool it as easily on Mars, as we can in the oceans of Earth. Radiation protection needs volume or distance. We can't produce nuclear fuel on Mars for a while and I don't even know, if there are meaningful amounts of Uranium near the surface to be mined. Swapping nuclear fuel on a rover is also something to give us some headaches. We can do that. Sure. But the realism would need to suffer to make it somewhat an option. At far more suffering for fun. And our settlers would need to become much smarter and far less suicidal.

Of course, things might be easier if the rover is unmanned. Then we could produce something like gigantic road trains that never get close enough to a base to matter. But these would require somewhat prepared roads and a better navigation system. No crew cabin would have the advantage, that the small reactor in front of that rover could be followed by a large radiator array, maybe even some on a trailer following the engine. Also the rover would no longer need to fit into a garage, only the trailer carts might have to go there. Still the economic problems stay, a very effective 1 MW reactor would need 25 kg of nuclear fuel replaced every year, but more realistic is a ballpark of 50 kg because of much the larger amount of neutron leakage. Since we can only fly every 18 months, we need something like 150 kg fuel rods (including spares if a lander goes AWOL) in the landers at each launch window. Can we transport this stuff on a regular lander, or do we need dedicated nuclear material transports? How much weight does the transport container add?

Like it is in the real world: Nuclear reactors are only cool while they are made of powerpoint slides, brochures and Youtube videos. HeXe coolant is a cool thing for a reactor in space, and really radiation resistant (compared to water), until you think about flow rates and the distance the coolant travels after a neutron capture and a few seconds of decays. And then the engineer goes back to the drawing board and ensures that astronauts do NOT get toasted when working on the supposingly safe side of the radiation shield... *sigh*

**Reply by mokun** [reply]
> because some media and nazi party told lies about a looming blackout.

Well these days media love doing "editorials", which need to be as sensational as possible to incite and appeal to viewer's emotion, throwing in a little bit distorted facts here and there, bending and twisting the fabric of reality to suit certain narratives to get a rich payback from sponsors...

> economic critic of nuclear power. Instead of maturing and getting cheaper and more reliable, the technology got more costly, has larger cost overruns additionally to it and its operators still manage to prepare the next large accident by accumulating technical debts. 

So we could conclude that technical or safety requirements for maintaining or managing nuclear plants hasn't received much breakthrough to drive down the cost.

> the fossile fuel consumption in the electricity sector dropped by nearly 40%, if you compare this years June to the past years)

So other than the imported energy from France, is fossil fuel still the largest source of power in Germany ?

**Reply by mokun** [reply]
> the fossile fuel consumption in the electricity sector dropped by nearly 40%, if you compare this years June to the past years)

I'd say it almost makes sense to create unmanned/automated network of harvester trains for transporting minerals to processing plants some days.

**Reply by mokun** [reply]
> then the engineer goes back to the drawing board and ensures that astronauts do NOT get toasted when working on the supposingly safe side of the radiation shield.

Well, if we want to be more sci-fi or exercise our right to believe in what is one day possible, we could go down the rabbit trail of creating a portable fusion power propulsion, quantum fluctuation/ZPE power box of some sort as our rover prototypes :D

**Comment by mokun** [comment]
So, what should be the top few major changes in 3.7.0 ?

**Reply by bevans2000** [reply]
Let's just capture the proposed changes as Issues and put them in the 3.7.0 Milestone. I've done some already.

**Comment by bevans2000** [comment]
But can you review the open issues and close some down that have been completed, e.g. #967

**Reply by mokun** [reply]
I just transferred 4 open issues from 3.1.0 to 3.7.0.

**Comment by mokun** [comment]
btw, v3.6.1 tag in GitHub has a problem.

![image](https://github.com/mars-sim/mars-sim/assets/1168584/3d3ab6b5-4cfe-4886-a2c5-c510216236f7)

- It should be `v3.6.1`, not `v.3.6.1`. 
- An extra dot right after the v is causing the tag to be unrecognized in README.md below :

![image](https://github.com/mars-sim/mars-sim/assets/1168584/a16b710b-b3a5-4766-8b2c-032bc93f480c)

Normally we should be able to change the name of a tag. 

It's probably a GitHub bug here. I've tried to change the tag name and the dot won't be allowed to go away.

**Reply by bevans2000** [reply]
That's my fault, I must have accidentally added the dot after the v. Not sure you can change the tag label. I will have a look and see

**Comment by bevans2000** [comment]
The simulation feels very fragile at the moment so for 3.7 we should focus on stability and performance improvements and not big new features.

**Reply by mokun** [reply]
> The simulation feels very fragile at the moment so for 3.7 

How about we create a `Develop` branch for incorporating those testy features ?

**Comment by bevans2000** [comment]
We could do but the master branch is the latest code base and has a lot of changes which seem to make it fragile. So making a new branch won't help in the short term. We need to make solid 3.7 release soon but focus on improvements. Then from a solid 3.7 we can adopt better technique and always use feature branches and test them before merging back to master.

**Reply by mokun** [reply]
>  has a lot of changes which seem to make it fragile

What do you see is being the most fragile part right now ?

To me, it seems we have just exposed a few buggy features that both 3.6.0 and 3.6.1 has been carrying. 

We have at least it attempted to fix some of them to hope to make them a little less fragile than already is. Do you think so ?

**Reply by bevans2000** [reply]
The change to using pending tasks instead of assigning directly as part of a Mission. It just seems to be causing problems like a lot of attempting to create Sleep and LoadVehicleTask that can not be started.
Plus it is slower as well

**Reply by mokun** [reply]
I just use visualvm to see what's consuming cpu cycles most : 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/1b160a9f-12aa-4836-8706-3230773eab9a)

The top 3 consumers are : 
1. 51.5 % - `Settlement::timePssingCitizens`
2. 23.9% - `Settlement::timePassing`
3. 15.8% - `BuildingManager::timePassing`

![image](https://github.com/mars-sim/mars-sim/assets/1168584/8e7de437-7ddf-4d13-9aed-16ae7f6a7462)

If I go inside `Settlement::timePssingCitizens`, two tasks are highlighted : 
1. 8.5% - `TendGreenhoue::tendingPhase()`
2. 11% - `DigLocal::performMappedPhase()`

**Reply by mokun** [reply]
I'll open an new issue to look into reducing the cpu cycles on them.

**Reply by bevans2000** [reply]
I gave noticed that the behaviour changes. From a few runs it starts ok then suddenly slows.

**Reply by mokun** [reply]
Documenting the slowness issue [here](https://github.com/mars-sim/mars-sim/issues/1039) for future tracking.

**Comment by mokun** [comment]
So, are we wrapping up 3.7.0 ? 

Anything else to include ? 

btw, we should also edit this [roadmap](https://github.com/mars-sim/mars-sim/wiki/Roadmap).

**Comment by bevans2000** [comment]
There are lots of open issues on the 3.7.0 milestone. I definitely want to complete #1084

**Reply by bevans2000** [reply]
Why don't you have a review of the Issues in 3.7.0 milestone and move any you think are not needed. I will review those assigned to me

**Reply by bevans2000** [reply]
However after the problems and long bugfix of 3.6 we need to stop any unnecessary changes once we have completed our goal issues. Allowing us to test the release and avoid last minute bugs.
I have moved quite a few 3.7.0 issues assigned to me. Those left I want to complete

**Reply by mokun** [reply]
I've just reviewed all the outstanding issues assigned to 3.7.0 previously.

Now they are down to 9 opens.

I'll work on wrapping primarily on 4 of them.

**Reply by bevans2000** [reply]
I've got time tomorrow to hopefully finish the bulk of #1084

**Reply by mokun** [reply]
We're down to only 6 open issues.

Which one(s) do you still want to work on or pushing them out to 3.8.0 ? 

https://github.com/mars-sim/mars-sim/milestones/3.7.0

**Comment by bevans2000** [comment]
#1029 for sure. I would like to have a look at #1119.
Also the dependabot PR are failing SonorQube. It's a known problem but I know how to fix it.
I've got some time at weekend to work on these.

**Comment by bevans2000** [comment]
Also we lack a lot on UnitTests around the config classes

**Reply by mokun** [reply]
> UnitTests around the config classes

Anything regarding the config class you may be worrying about?

**Reply by bevans2000** [reply]
Not really but coverage is low and these classes are large and complex.

**Comment by mokun** [comment]
Now we're down to these last 3 open issues : 
1. https://github.com/mars-sim/mars-sim/issues/1143
2. https://github.com/mars-sim/mars-sim/issues/1119
3. https://github.com/mars-sim/mars-sim/issues/954

Do we want more work on them to close them out or push them out ?

**Reply by bevans2000** [reply]
There are big changes in 1143 and last time that broke the release at the last minute. So the master branch needs time to be tested and some simulation runs done with out changes to ensure it is stable.
 I can live with 1119 going to the next release. I know nothing about 954.

**Reply by mokun** [reply]
With all the great things we've done after 3.6.2, can you summarize some in `whatsnew.html` in `mars-sim-ui` ? 

We're going to wrap up and push out 3.7.0 very soon, right ?

**Reply by bevans2000** [reply]
Yes I agree. Do you want me to push the work on #47 ?

**Reply by mokun** [reply]
No they're good to go to be included in 3.7.0.

**Reply by bevans2000** [reply]
So we include #47 changes. Shall I create a pull request tomorrow then

**Reply by mokun** [reply]
Sure. 

Anything other important things to wrap up ?

**Comment by mokun** [comment]
> There are big changes in 1143 and last time that broke the release at the last minute. So the master branch needs time to be tested and some simulation runs done with out changes to ensure it is stable.

It's stable now.

**Comment by bevans2000** [comment]
We need to consider moving to protected branch for master branch. Its common these days when there is more than one developer.

**Comment by mokun** [comment]
Unless you have something you're working on to incorporate into 3.7.0, let's do finalize things and package it for the release. Agree ?

**Comment by mokun** [comment]
Can you add to the change log in whatsnew.html ?

**Reply by bevans2000** [reply]
Yes let's close it down. I will add to the what's new when I next logon

**Comment by mokun** [comment]
Any other important bug fixes for v3.7.1, other than time execution spike ?

**Reply by bevans2000** [reply]
No

**Reply by mokun** [reply]
We've just released v3.7.1.

But how do I update its docker image ? 

[Here](https://github.com/mars-sim/mars-sim/pkgs/container/mars-sim%2Fmars-sim) it still shows v3.7.0.

**Reply by mokun** [reply]
Why are there so many unnamed images (with hashes and no real names) at https://github.com/mars-sim/mars-sim/pkgs/container/mars-sim%2Fmars-sim/versions ? 

Can you delete them ?

**Reply by bevans2000** [reply]
You manually run the Docker Action against the tag. I have just started it.
Yes I can remove the unnamed packages

**Reply by mokun** [reply]
> Yes I can remove the unnamed packages

I just deleted a few one by one [here](https://github.com/mars-sim/mars-sim/pkgs/container/mars-sim%2Fmars-sim/versions). 

But there are still 1199 of them there were not tagged. Is there a quick way to select them and click delete ?

**Reply by mokun** [reply]
> You manually run the Docker Action against the tag. 

Did you run it on this [page](https://github.com/mars-sim/mars-sim/actions/workflows/docker-publish.yml) ?

**Reply by bevans2000** [reply]
Yes

**Comment by bevans2000** [comment]
> > Yes I can remove the unnamed packages
> 
> I just deleted a few one by one [here](https://github.com/mars-sim/mars-sim/pkgs/container/mars-sim%2Fmars-sim/versions).
> 
> But there are still 1199 of them there were not tagged. Is there a quick way to select them and click delete ?

After a bit of research seems the best way is to create a GitHub Action that can purge untagged packages. I'll have a look at weekend

**Reply by mokun** [reply]
Or we may go to GitHub's Issue page and file a feature request to have them add the option of deleting all the old docker images.

**Reply by bevans2000** [reply]
You could but it's probably not an ùrgent feature if they do accept it.

**Reply by bevans2000** [reply]
I have created a new GitHub Action called 'Remove Untagged Packages" that deleted them. But It can only do 100 at a time so I have run it a few times to clean up.

**Reply by mokun** [reply]
Are we done with what we have wanted to do in 3.7.x ?

**Reply by bevans2000** [reply]
Yes
